aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp

```diff
@@ -0,0 +1,116 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cpu/Loops.h>
+#include <ATen/native/quantized/cpu/quantized_ops.h>
+#include <ATen/quantized/Quantizer.h>
+#include <ATen/native/quantized/cpu/init_qnnpack.h>
+#include <ATen/native/quantized/cpu/qnnpack_utils.h>
+#include <c10/core/TensorOptions.h>
+#include <caffe2/utils/threadpool/ThreadPoolMobile.h>
+
+#include <algorithm>
+
+namespace at {
+namespace native {
+
+namespace {
+Tensor quantized_channel_shuffle_impl(
+    const Tensor& self,
+    int64_t groups) {
+
+  TORCH_CHECK(
+      self.dim() == 4,
+      "channel_shuffle expects 4D input, but got input with sizes ",
+      self.sizes());
+  TORCH_CHECK(
+      self.scalar_type() == kQUInt8,
+      "Quantized channel shuffle works only on uint8_t.",
+      "But got:", self.scalar_type());
+  const Tensor self_nhwc = self.contiguous(MemoryFormat::ChannelsLast);
+  Tensor qy = at::new_qtensor_cpu(
+      self_nhwc.sizes(),
+      TensorOptions(kQUInt8)
+          .device(kCPU)
+          .memory_format(self_nhwc.suggest_memory_format()),
+      make_per_tensor_affine_quantizer(
+          self_nhwc.q_scale(), self_nhwc.q_zero_point(), kQUInt8)
+      );
+
+  // Degenerate case of just copying.
+  if (groups == 1) {
+    qy.copy_(self_nhwc);
+    return qy.contiguous(self.suggest_memory_format());
+  }
+
+  int64_t channels = self.size(1);
+  TORCH_CHECK((channels % groups) == 0,
+             "Number of channels must be divisible gy groups. Got ",
+             channels, " channels and ", groups, " groups.");
+
+  initQNNPACK();
+
+  pytorch_qnnp_operator_t qnnpack_operator{nullptr};
+
+  const pytorch_qnnp_status createStatus = pytorch_qnnp_create_channel_shuffle_nc_x8(
+      groups /* groups */,
+      channels / groups /* group channels */,
+      0 /* flags */,
+      &qnnpack_operator);
+  TORCH_INTERNAL_ASSERT(
+      createStatus == pytorch_qnnp_status_success,
+      "failed to create QNNPACK Add operator");
+
+  const pytorch_qnnp_status setupStatus = pytorch_qnnp_setup_channel_shuffle_nc_x8(
+      qnnpack_operator,
+      self_nhwc.numel() / channels /* batch size */,
+      (uint8_t*)self_nhwc.data_ptr<c10::quint8>() /* slef data */,
+      channels /* self stride */,
+      (uint8_t*)qy.data_ptr<c10::quint8>() /* qy data */,
+      channels /* qy stride */);
+  TORCH_INTERNAL_ASSERT(
+      setupStatus == pytorch_qnnp_status_success,
+      "failed to setup QNNPACK Add operator");
+
+#ifdef FBCODE_CAFFE2
```

Comment:
Can we make mobile_pthreadpool return null in fbcode instead of checking here?

Reply:
Most of this function was just lifted from C2's implementation. However I think it is cleaner to read this code. However, maybe we should actually pass a valid threadpool. The one that at::parallel_for etc. uses.

Reply:
I don't think this ifdef is necessary. Please check other files like qadd.cpp

Reply:
OK, looks like to other implementations, such as qadd etc dont have this ifdef. I still think it would have been cleaner if to use mobile_threadpool only for mobile. Now we have code where we are using mobile threadpool even when we may not be running on mobile usecase. It makes this little difficult to read. But for consistency sake, I will stick with the convention.